### PR TITLE
first implementation yarn support

### PR DESCRIPTION
**Pull request by maybeec**
This is a PR for #490.
It just provides the first support for installing and running yarn. It has not yet been testing extensively, but embedded in a local build. It has been tested for windows only until now.

### CODE REVIEW

**Comment by eirslett** [comment]
Looks promising! Could you please add an integration test that installs node and yarn, then installs a couple of libs and runs some kind of test? Like the example project?

**Comment by maybeec** [comment]
Yes, would be nice. I may have a look at this at the weekend if possible.

**Comment by mriehema** [comment]
changelog and documentation is missing too.

**Comment by maybeec** [comment]
where should I document anything? The existing code is missing documentation as well.

**Comment by mriehema** [comment]
a new section in README.md to describe how to configure pom for yarn (should be similar to install-node-and-npm) and a simple one-liner in CHANGELOG.md should be enough for beginning.

**Comment by eirslett** [comment]
The existing readme just points to the example project/integration tests - that way, the code is self-documenting and the documentation is always up-to-date. (At least in theory)

File: frontend-maven-plugin/src/it/yarn-integration/verify.groovy [hunk 1]
```diff
@@ -0,0 +1,8 @@
+assert new File(basedir, 'target/.buildenv/node').exists() : "Node was not installed in the custom install directory";
+assert new File(basedir, 'node_modules').exists() : "Node modules were not installed in the base directory";
+
+import org.codehaus.plexus.util.FileUtils;
+
+String buildLog = FileUtils.fileRead(new File(basedir, 'build.log'));
+
+assert buildLog.contains('BUILD SUCCESS') : 'build was not successful'
```
Comment (eirslett) [inline frontend-maven-plugin/src/it/yarn-integration/verify.groovy]: Maybe also assert that node_modules/less/package.json exists? Then we're sure the plugin has actually installed dependencies, not just short-circuited!

**Comment by maybeec** [comment]
finally, got it to work also on linux. However, the download URL is just based on GitHub releases of Yarn. I did not found any webroot with all deployments. Hope, that the yarn guys will keep their naming conventions of the binaries consistent :)

I also adapted the test accordingly.

File: frontend-maven-plugin/src/it/custom-install-directory/verify.groovy [hunk 2]
```diff
@@ -1,2 +1,2 @@
-assert new File(basedir, 'target/node').exists() : "Node was not installed in the custom install directory";
+assert new File(basedir, 'target/.buildenv/node').exists() : "Node was not installed in the custom install directory";
```
Comment (eirslett) [inline frontend-maven-plugin/src/it/custom-install-directory/verify.groovy]: I'm not sure it's such a good idea to add the .buildenv/ directory, since it will change the directory layout, which could hurt backwards compatibility. Maybe we can just have node/node(.exe), node/npm/xxx and node/yarn/xxx?

File: frontend-maven-plugin/src/main/java/com/github/eirslett/maven/plugins/frontend/mojo/InstallNodeAndNpmMojo.java [hunk 3]
```diff
@@ -73,40 +79,38 @@ protected boolean skipExecution() {
...
         if (null != server) {
-            factory.getNodeAndNPMInstaller(proxyConfig)
-                .setNodeVersion(nodeVersion)
-                .setNodeDownloadRoot(nodeDownloadRoot)
-                .setNpmVersion(npmVersion)
-                .setNpmDownloadRoot(npmDownloadRoot)
-                .setUserName(server.getUsername())
-                .setPassword(server.getPassword())
-                .install();
+            factory.getNodeInstaller(proxyConfig).setNodeVersion(this.nodeVersion)
```
Comment (eirslett) [inline frontend-maven-plugin/src/main/java/com/github/eirslett/maven/plugins/frontend/mojo/InstallNodeAndNpmMojo.java]: Why has this stuff been changed? It's not necessary.

File: frontend-maven-plugin/src/main/java/com/github/eirslett/maven/plugins/frontend/mojo/InstallNodeAndYarnMojo.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        return this.skip;
+    }
+
+    @Override
+    public void execute(FrontendPluginFactory factory) throws InstallationException {
+        ProxyConfig proxyConfig = MojoUtils.getProxyConfig(this.session, this.decrypter);
+        Server server = MojoUtils.decryptServer(this.serverId, this.session, this.decrypter);
+        if (null != server) {
+            factory.getNodeInstaller(proxyConfig).setNodeDownloadRoot(this.nodeDownloadRoot)
+                .setNodeVersion(this.nodeVersion).setPassword(server.getPassword())
```
Comment (eirslett) [inline frontend-maven-plugin/src/main/java/com/github/eirslett/maven/plugins/frontend/mojo/InstallNodeAndYarnMojo.java]: It's much easier to read if there is one method call per line.

File: frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/ArchiveExtractor.java [hunk 5]
```diff
@@ -29,86 +34,101 @@
...
     public void extract(String archive, String destinationDirectory) throws ArchiveExtractionException {
         final File archiveFile = new File(archive);
-        FileInputStream fis = null;
-        try {
-            fis = new FileInputStream(archiveFile);
 
-            if("zip".equals(FileUtils.getExtension(archiveFile.getAbsolutePath()))){
+        try (FileInputStream fis = new FileInputStream(archiveFile)) {
+            if ("msi".equals(FileUtils.getExtension(archiveFile.getAbsolutePath()))) {
+                String command = "msiexec /a " + archiveFile.getAbsolutePath() + " /qn TARGETDIR=\""
```
Comment (eirslett) [inline frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/ArchiveExtractor.java]: Will this cause the yarn installation to be registered globally in the system, even if the files are placed in the local folder? The operating system shouldn't know about yarn being installed (because then it can be accidentally uninstalled).

File: frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/FrontendPluginFactory.java [hunk 6]
```diff
@@ -3,34 +3,45 @@
 import java.io.File;
 
 public final class FrontendPluginFactory {
-    
+
```
Comment (eirslett) [inline frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/FrontendPluginFactory.java]: It's not necessary to fix all code style things here, adding new methods is enough. (I suppose code style fixes could be a separate PR)

File: frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/NodeExecutorConfig.java [hunk 7]
```diff
@@ -3,43 +3,48 @@
...
+    @Override
+    public File getNpmPath() {
+        return new File(this.installConfig.getInstallDirectory() + Utils.normalize(NPM));
+    }
 
-  @Override
-  public File getWorkingDirectory() {
-    return installConfig.getWorkingDirectory();
-  }
+    @Override
```
Comment (eirslett) [inline frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/NodeExecutorConfig.java]: Fixing the code style can be done in a separate PR (to keep the line change count down on this PR)

File: frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/YarnTaskExecutor.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+            final boolean maskMavenProxyPassword = s.contains("proxy=");
+            if (maskMavenProxyPassword) {
+                final String bestEffortMaskedPassword = maskPassword(s);
+                clonedArguments.set(i, bestEffortMaskedPassword);
+            }
+        }
+        return "'" + taskName + " " + implode(" ", clonedArguments) + "'";
+    }
+
+    private static String maskPassword(String proxyString) {
```
Comment (eirslett) [inline frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/YarnTaskExecutor.java]: I guess this function could be moved into a separate helper class somewhere? I suspect it's used somewhere else too.

File: frontend-maven-plugin/src/it/custom-install-directory/verify.groovy (hunk 2 above)
Comment (maybeec) [inline frontend-maven-plugin/src/it/custom-install-directory/verify.groovy]: basically, it is your decision. However, the name of the temporary directory `node` is not a good choice especially, if you want to ignore it on commit. Therefore, I introduced a more speaking and generic folder name to also include further tmp local environments in future.
I even do not see any "backward-compatibility issues" here as this is just a folder of temporary files, which will even be recreated at build time, but correct me if I am wrong.

So I would highly recommend to change the folder name.

File: frontend-maven-plugin/src/main/java/com/github/eirslett/maven/plugins/frontend/mojo/InstallNodeAndNpmMojo.java (hunk 3 above)
Comment (maybeec) [inline frontend-maven-plugin/src/main/java/com/github/eirslett/maven/plugins/frontend/mojo/InstallNodeAndNpmMojo.java]: It has just been changed due to IDE auto-formatting reasons. Basically, I was lazy to format code by myself. However, I would guess that you have your own favorite code style, which is not shared anyhow? I am applying my code styles while merging before doing a review. Thus, I can see what actually has been changed and merge the code with the correct code style.

File: frontend-maven-plugin/src/main/java/com/github/eirslett/maven/plugins/frontend/mojo/InstallNodeAndYarnMojo.java (hunk 4 above)
Comment (maybeec) [inline frontend-maven-plugin/src/main/java/com/github/eirslett/maven/plugins/frontend/mojo/InstallNodeAndYarnMojo.java]: personal preference, but here just as of my used IDE settings.

File: frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/ArchiveExtractor.java (hunk 5 above)
Comment (maybeec) [inline frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/ArchiveExtractor.java]: As far as I got to know, the msiexec command just extracts the archive to a given target folder. There is no software registry updated.

File: frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/YarnTaskExecutor.java (hunk 8 above)
Comment (eirslett) [inline frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/YarnTaskExecutor.java]: Yes, it's right that the visibility stuff is written a bit weird, it should probably be changed so all classes are visible, at least in their own package. I suppose we can skip refactoring this one then, and to it later :-)

File: frontend-maven-plugin/src/it/custom-install-directory/verify.groovy (hunk 2 above)
Comment (mriehema) [inline frontend-maven-plugin/src/it/custom-install-directory/verify.groovy]: Please add a separate PR for this change to discuss. This has nothing to do with yarn and would break a lot of existing projects.

File: frontend-maven-plugin/src/main/java/com/github/eirslett/maven/plugins/frontend/mojo/InstallNodeAndNpmMojo.java [hunk 9]
```diff
@@ -1,45 +1,51 @@
...
 import org.apache.maven.plugins.annotations.Parameter;
-import org.apache.maven.settings.crypto.SettingsDecrypter;
 import org.apache.maven.settings.Server;
+import org.apache.maven.settings.crypto.SettingsDecrypter;
+
+import com.github.eirslett.maven.plugins.frontend.lib.FrontendPluginFactory;
+import com.github.eirslett.maven.plugins.frontend.lib.InstallationException;
+import com.github.eirslett.maven.plugins.frontend.lib.NPMInstaller;
+import com.github.eirslett.maven.plugins.frontend.lib.NodeInstaller;
+import com.github.eirslett.maven.plugins.frontend.lib.ProxyConfig;
```
Comment (mriehema) [inline frontend-maven-plugin/src/main/java/com/github/eirslett/maven/plugins/frontend/mojo/InstallNodeAndNpmMojo.java]: Please revert all changes, which are not needed for yarn. That would help us to review your PR.

File: frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/ArchiveExtractor.java (hunk 5 above)
Comment (mriehema) [inline frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/ArchiveExtractor.java]: I will test that on my Windows machine.

**Comment by mandragorn** [comment]
Is there anything I can do to help out with this?  I'm actively looking at switching our organization over to yarn and we currently use the frontend-maven-plugin.  Until this is merged I'll be using the workarounds described in #490

**Comment by maybeec** [comment]
Actually, the only open issue here is the code formatting as far as I can see? Right?
As I am currently running out of time, I will not be able to rework the code, I can give anyone access to my fork if interested. However, I would recommend to the plugin authors to do it by their own if they really want to do so. 
By the way: From my experiences, GitHub UI seems to not be able to calculate the final diff of the PR and the master branch. It will moreover create a view by just aggregating all changes, which leads you to review changes, which have been eliminated by later commits on the PR. So I would recommend to just merge and review on the local machine.
Then review might be even easier.

**Comment by maybeec** [comment]
@mriehema can you argument on your comment? What should break for already existing projects, when using a different tmp folder?
I would even argument against it and say, that the current folder "node" is a bad choice regarding name. The name does not include any of its original semantics and it might be even worse when trying to exclude it recursively from version control systems as "node" might be an often used folder name in code.

**Comment by maiis** [comment]
Looks awesome. Any plan to merge this feature ? Can't wait 👍 💯

**Comment by eirslett** [comment]
There are some concerns that have to be addressed before it is merged (read the thread), I suspect @mandragorn or @mriehema are willing to take a look at them?

**Comment by polarctos** [comment]
I addressed all three concerns in https://github.com/polarctos/frontend-maven-plugin/commit/c2c053323cd7dc68ff714d4509cbe0d76bdb665d:
* reverted the formatting changes
* removed usage of the .buildenv folder
* added a section to the readme file

Is there anything else that is missing to get this squashed and merged?

**Comment by eirslett** [comment]
Thanks a lot! That should be okey then! Can you squash everything into one commit and make a new PR to this repo? Then I'll give it another review and merge it finally :-)

**Comment by eirslett** [comment]
This PR was updated in #499 and merged, so I'll close this issue.
